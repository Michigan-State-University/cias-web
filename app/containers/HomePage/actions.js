import { actionBuilder } from 'utils/actionBuilder';

import {
  CREATE_INTERVENTION_REQUEST,
  CREATE_INTERVENTION_SUCCESS,
  CREATE_INTERVENTION_ERROR,
} from './constants';

export const createInterventionRequest = () =>
  actionBuilder(CREATE_INTERVENTION_REQUEST, {});
export const createInterventionSuccess = () =>
  actionBuilder(CREATE_INTERVENTION_SUCCESS, {});
export const createInterventionError = error =>
  actionBuilder(CREATE_INTERVENTION_ERROR, { error });
