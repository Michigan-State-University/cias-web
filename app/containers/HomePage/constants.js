export const CREATE_INTERVENTION_REQUEST =
  'app/Dashboard/CREATE_INTERVENTION_REQUEST';
export const CREATE_INTERVENTION_SUCCESS =
  'app/Dashboard/CREATE_INTERVENTION_SUCCESS';
export const CREATE_INTERVENTION_ERROR =
  'app/Dashboard/CREATE_INTERVENTION_ERROR';
