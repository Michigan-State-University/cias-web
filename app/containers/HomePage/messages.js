/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.containers.HomePage';

export default defineMessages({
  createIntervation: {
    id: `${scope}.createIntervation`,
    defaultMessage: 'Create intervation',
  },
});
