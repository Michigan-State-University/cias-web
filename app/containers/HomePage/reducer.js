import produce from 'immer';

import {
  CREATE_INTERVENTION_REQUEST,
  CREATE_INTERVENTION_ERROR,
  CREATE_INTERVENTION_SUCCESS,
} from './constants';

export const initialState = {
  loaders: {
    interventionCreating: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const dashboardPageReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CREATE_INTERVENTION_REQUEST:
        draft.interventionCreating = true;
        break;

      case CREATE_INTERVENTION_SUCCESS:
        draft.interventionCreating = false;
        break;

      case CREATE_INTERVENTION_ERROR:
        draft.interventionCreating = false;
        break;
    }
  });

export default dashboardPageReducer;
