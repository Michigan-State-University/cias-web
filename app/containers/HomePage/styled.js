import styled from 'styled-components';
import { themeColors, colors } from 'theme';

export const HomePageContainer = styled.div``;

export const NewInterventionFloatButton = styled.div`
  position: fixed;
  right: 115px;
  cursor: pointer;
  bottom: 20px;
  background: ${themeColors.secondary};
  color: white;
  border-radius: 5px;
  padding: 15px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const AddIcon = styled.div`
  width: 21px;
  height: 21px;
  border-radius: 50%;
  background: ${colors.white};
  display: inline-block;
  color: ${themeColors.secondary};
  font-size: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: 5px;
`;
