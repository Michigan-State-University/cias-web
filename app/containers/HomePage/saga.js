import { put, takeLatest, all } from 'redux-saga/effects';
import axios from 'axios';
import { push } from 'connected-react-router';

import { CREATE_INTERVENTION_REQUEST } from './constants';

import { createInterventionSuccess, createInterventionError } from './actions';

function* createIntervention() {
  const requestURL = `v1/interventions`;

  try {
    const response = yield axios.post(requestURL, {
      intervention: {
        type: 'Intervention::Single',
        name: 'e-Intervention New',
      },
    });

    yield put(createInterventionSuccess());
    yield put(push(`/interventions/${response.data.data.id}/edit`));
  } catch (error) {
    yield put(createInterventionError(error));
  }
}

export default function* dashboardPageSaga() {
  yield all([
    yield takeLatest(CREATE_INTERVENTION_REQUEST, createIntervention),
  ]);
}
