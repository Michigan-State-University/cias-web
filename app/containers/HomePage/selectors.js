import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectDashboardPageDomain = state => state.dashboardPage || initialState;

export const makeSelectDashboardPage = () =>
  createSelector(
    selectDashboardPageDomain,
    substate => substate,
  );

export const makeSelectInterventionCreatingLoader = () =>
  createSelector(
    selectDashboardPageDomain,
    substate => substate.loaders.interventionCreating,
  );
