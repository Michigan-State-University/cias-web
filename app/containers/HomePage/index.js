/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React, { useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { Container, Row, Col } from 'react-grid-system';

import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';

import Loader from 'components/Loader';
import ErrorAlert from 'components/ErrorAlert';
import H1 from 'components/H1';
import SingleInterventionPanel from 'components/SingleInterventionPanel';
import Spinner from 'components/Spinner';
import {
  interventionListReducer,
  fetchInterventionsSaga,
  makeSelectInterventionList,
  fetchInterventionsRequest,
} from 'global/reducers/interventionList';

import {
  createInterventionSaga,
  createInterventionRequest,
  makeSelectInterventionLoaders,
} from 'global/reducers/intervention';

import { NewInterventionFloatButton, AddIcon } from './styled';
import messages from './messages';
export function HomePage({
  createIntervention,
  fetchInterventions,
  loaders: { createIntervention: createInterventionLoader },
  interventionList: {
    interventions,
    fetchInterventionLoading,
    fetchInterventionError,
  },
}) {
  useInjectSaga({ key: 'createIntervention', saga: createInterventionSaga });
  useInjectReducer({
    key: 'interventionList',
    reducer: interventionListReducer,
  });
  useInjectSaga({ key: 'getInterventions', saga: fetchInterventionsSaga });

  useEffect(() => {
    fetchInterventions();
  }, []);

  const wrapWithCol = (child, key) => (
    <Col key={`Single-intvention-${key}`} xs={12} sm={6} lg={4} xl={3}>
      {child}
    </Col>
  );

  if (fetchInterventionLoading) return <Loader />;
  if (fetchInterventionError)
    return (
      <Container>
        <ErrorAlert errorText={fetchInterventionError} />
      </Container>
    );

  return (
    <Fragment>
      <Container>
        <H1 my={35}>
          <FormattedMessage {...messages.myIntervention} />
        </H1>
        <Row>
          {wrapWithCol(
            <SingleInterventionPanel
              clickHandler={createIntervention}
              interventionCreating={createInterventionLoader}
            />,
            'new',
          )}
          {interventions.map(intervention =>
            wrapWithCol(
              <SingleInterventionPanel intervention={intervention} />,
              intervention.id,
            ),
          )}
        </Row>
      </Container>
      <NewInterventionFloatButton onClick={createIntervention}>
        {!createInterventionLoader && (
          <>
            <AddIcon>+</AddIcon>
            <FormattedMessage {...messages.createIntervention} />
          </>
        )}
        {createInterventionLoader && <Spinner />}
      </NewInterventionFloatButton>
    </Fragment>
  );
}

HomePage.propTypes = {
  intl: PropTypes.object,
  createIntervention: PropTypes.func,
  fetchInterventions: PropTypes.func,
  homePageState: PropTypes.object,
  interventionList: PropTypes.object,
  loaders: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  interventionList: makeSelectInterventionList(),
  loaders: makeSelectInterventionLoaders(),
});

const mapDispatchToProps = {
  createIntervention: createInterventionRequest,
  fetchInterventions: fetchInterventionsRequest,
};

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(HomePage);
