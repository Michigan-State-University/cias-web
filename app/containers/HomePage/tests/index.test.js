import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';

import { ConnectedRouter } from 'connected-react-router';
import { Provider } from 'react-redux';
import configureStore from 'configureStore';
import { browserHistory } from 'react-router-dom';
import HomePage from '../index';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it.skip('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale="en">
        <Provider store={store}>
          <ConnectedRouter>
            <HomePage />
          </ConnectedRouter>
        </Provider>
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
