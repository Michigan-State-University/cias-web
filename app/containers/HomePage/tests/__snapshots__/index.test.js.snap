// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div
  class="styled__HomePageContainer-r8av8f-0 hPMtJa"
>
  <button
    class="StyledButton-t20ntd-0 goZETI"
    color="primary"
    title="Create intervention"
    width="100%"
  >
    Create intervention
  </button>
</div>
`;
